### DISCUSSION TITLE
Weird types in middleware

### DISCUSSION THREAD

**Discussion by MalekD5** [socketio/socket.io#4948] [Q&A] [answered]
When I try to use middleware `socket.use()`, for some reason it returns the incorrect types

![image](https://github.com/socketio/socket.io/assets/106095300/401244fc-672b-47dc-be46-f3a7cbc61f81)

when I log the `req` parameter, it returns a `Socket` object

![image](https://github.com/socketio/socket.io/assets/106095300/4279e15b-6c78-4217-a16e-24e4688917f3)

I'm not sure if this is a result from my typescript setup. but I have to do this so typescript stops complaining
![image](https://github.com/socketio/socket.io/assets/106095300/a22d224b-fe37-4689-8e39-520d26b24ec7)

**Comment by darrachequesne** [comment]
Hi! Which kind of middleware are you using?

- at the namespace level?

```js
io.use((socket, next) => {
  // ...
});
```

Reference: https://socket.io/docs/v4/server-api/#serverusefn

- or at the socket level?

```js
io.on("connection", (socket) => {
  socket.use(([event, ...args], next) => {
    // ....
  });
});
```

Reference: https://socket.io/docs/v4/server-api/#socketusefn

**Reply by MalekD5** [reply, marked as answer]
Hi, as it turns out I was incorrectly inferring the types by using `Socket` type instead of `Server` type. This confusion was related to nestjs websocket package that was using the type `any` for the lifecycle hooks ([`afterInit`](https://github.com/nestjs/nest/blob/3e61ec1bea126e6881be126e6d3941db33ac594e/packages/websockets/interfaces/nest-gateway.interface.ts#L5) to be precise) not realizing that the parameter was referring to a server rather than a client (Socket).

I think I should have been asking for support directly from nestjs rather than socket.io 
My apologies.
